miteD/middleware/api.py

```diff
@@ -42,10 +70,52 @@ def stop(self):
                     with suppress(asyncio.CancelledError):
                         self._loop.run_until_complete(task)
                 self._loop.close()
+                self._deregister_with_consul()
 
             def get_remote_service(self, service_name, version):
                 return RemoteService(name=service_name, version=version, nc=self._nc)
 
+            def _register_with_consul(self):
+                self.consul_id = os.getenv("HOSTNAME")
+                # This is also the pod name
+
+                api_data = {
+                  "Name": self.consul_id,
+                  "ID": self.consul_id,
+                  "Address": f"api/{name}/{versions[0]}",
+                  "Tags": [
+                      "api",
+                      f"{versions[0]}",
+                      f"{name}"
+                  ],
+                  "Service": {
+                    "ID": self.consul_id,
+                    "Service": "chip-api"
+                  },
+                  "Check": {
+                      "CheckID": f"{self.consul_id}-TTLCheck",
+                      "DeregisterCriticalServiceAfter": KEEP_DEGRADED_SERVICE,
+                      "TTL": TTL_TIMEOUT_INTERVAL
+                  }
+                }
+
+                r = requests.put(
+                    CONSUL_ADDRESS + "/v1/agent/service/register",
+                    data=json.dumps(api_data)
+                )
+
+            def _deregister_with_consul(self):
+                requests.put(CONSUL_ADDRESS + f"/v1/agent/service/deregister/{self.consul_id}")
```

Comment:
what if this fails? Is it detrimental to our service that we cannot deregister?

Reply:
Consul would think the services was failing checks rather than intentionally removed and would list it as failing until `KEEP_DEGRADED_SERVICE` variable has elapsed at which point it would be removed from the registry. 

As far as the middleware/api itself. It shouldn't make a difference. This executes as the pod is shutting down anyways. I've tried it with and without the Consul service registry being up.

Reply:
sounds good, thanks for the elaboration